flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/WaitingForFirstBarrierUnaligned.java

```diff
@@ -63,14 +63,18 @@ public BarrierHandlerAction barrierReceived(
         CheckpointBarrier unalignedBarrier = checkpointBarrier.asUnaligned();
         context.triggerTaskCheckpoint(unalignedBarrier);
         for (CheckpointableInput input : inputs) {
-            input.checkpointStarted(checkpointBarrier);
+            input.checkpointStarted(unalignedBarrier);
         }
         context.triggerGlobalCheckpoint(unalignedBarrier);
         if (context.allBarriersReceived()) {
             for (CheckpointableInput input : inputs) {
-                input.checkpointStopped(checkpointBarrier.getId());
+                input.checkpointStopped(unalignedBarrier.getId());
+            }
+            if (alternating) {
+                return new AlternatingWaitingForFirstBarrier(inputs);
+            } else {
+                return this;
```

Comment:
When would we not be alternating?

Reply:
So far only in tests. See `UnalignedCheckpointsTest`.